WIP(core): add test util methods to fakeAsync()

JiaLiPassion opened the pull request:
Add some util methods for `fakeAsync()` inspired by `jest`.  https://jestjs.io/docs/en/jest-object#mock-timers

Since `zone.js` now support integration with `jest fakeTimers`, internally zone.s `FakeAsyncZoneSpec` already supports all these functionalities of `jest fakeTimers`, so this PR expose those abilities as public APIs.

1. `getFakeBaseSystemTime()`.

 In `fakeAsync()`, the `Date.now()` is patched, so it
 always return a fixed number unless calling `tick()` to advance
 the virtual timer. This API returns the current fake system time.
 It will be the same with calling `Date.now()` in the `fakeAsync()`

2. `setFakeBaseSystemTime(time)`

 In `fakeAsync()`, the `Date.now()` is patched, so it
 always return a fixed number unless calling `tick()` to advance
 the virtual timer. This API set the current fake system time.

3. `getRealSystemTime()`

 In `fakeAsync()`, the `Date.now()` is patched, so it
 always return a fixed number unless calling `tick()` to advance
 the virtual timer. This API get the underlying real system time(unpatched)

4. `flushOnlyPendingTimers()`.

Flush all the pending timers, if any new timers are spawn,
 those new timers will not be consumed.

For example:
```
it('test nested timeout`, fakeAsync(() => {
  const logs = [];
  setTimeout(() => {
    logs.push(1);
    setTimeout(() => {
      logs.push(2);
    });
  });
  // flush(); will invoke the nested timeout as well, logs will be [1,2]
 // flushOnlyPendingTimeout will only flush the current pending timeout, the logs will be [1]
}));
```

5. `tickToNext(steps)`

tick to the `steps` amount of timers, steps by default is 1.

```
it('test nested timeout`, fakeAsync(() => {
  const logs = [];
  setTimeout(() => {
    logs.push(1);
  }, 100);
  setTimeout(() => {
    logs.push(11);
  }, 100);
  setTimeout(() => {
    logs.push(2);
  }, 200);
  setTimeout(() => {
    logs.push(3);
  }, 300);

  tickToNext();
  expect(logs).toEqual([1,11]); 
  tickToNext(2);
  expect(logs).toEqual([1, 11, 2, 3]);
}));
```

6. `removeAllTimers()`

Remove all the pending timers, intervals, microTasks.

7. `getTimerCount()`

 Get the count of the pending timers, intervals and microTasks.

mhevery reviewed (commented):
These new APIs seem reasonable to me. Can you work with @kapunahelewong on making sure that the API documentation is upto Angular standard. Thanks.

jelbourn commented:
Let's discuss the APIs at the team meeting before diving into the docs

JiaLiPassion commented:
@mhevery , @jelbourn , got it, thanks!

IgorMinar commented on packages/zone.js/lib/testing/fake-async.ts:
```diff
@@ -137,14 +137,85 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
+  /**
+   * set fake system time, in millis.
+   *
+   * @experimental
```
it's unusual that stable Angular APIs depend on experimental zone.js apis. Is that intentional?

JiaLiPassion replied:
Yeah, I think we can remove the `experimental` now.

IgorMinar commented on packages/zone.js/lib/testing/fake-async.ts:
```diff
@@ -137,14 +137,85 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
+  /**
+   * tick to the next timer.
+   *
+   * @param steps, the steps tick to, by default is 1, the steps parameter can not < 1.
```
```suggestion
   * @param steps, the steps tick to, by default is 1, the steps parameter can not be less than 1.
```

JiaLiPassion replied:
Got it, thanks.

IgorMinar commented on packages/zone.js/lib/testing/fake-async.ts:
```diff
@@ -137,14 +137,85 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
+  }
+
+  /**
+   * get fake system time, in millis.
```
```suggestion
   * Get fake system time, in millis.
```

please capitalize all the other ones as well.

JiaLiPassion replied:
Got it, thanks.

IgorMinar commented on packages/zone.js/lib/testing/fake-async.ts:
```diff
@@ -137,14 +137,85 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
+  }
+
+  /**
+   * set underlying native system time, in millis.
```
this seems like a comment that doesn't relate to this function. Can you please take a look?

JiaLiPassion replied:
Got it, thanks. This is a mistake.

IgorMinar commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -156,3 +156,129 @@ export function flushMicrotasks(): void {
+/**
+ * Flush only pending timers.
+ *
+ * Flush all the pending timers, if any new timers are spawn,
```
Do you mean that if they are spawned during the flushing of currently pending timers? Essentially meaning that we don't "recursively" flush the timers and we don't guarantee that the system is in a stable (no pending timers) state.

JiaLiPassion replied:
So the case look like this 
```
it('test nested timeout`, fakeAsync(() => {
  const logs = [];
  setTimeout(() => {
    logs.push(1);
    setTimeout(() => {
      logs.push(2);
    });
  });
  // flush(); will invoke the nested timeout as well, logs will be [1,2]
 // flushOnlyPendingTimeout will only flush the current pending timeout, the logs will be [1]
}));
```

So this new `flushOnlyPendingTimeout` will only take care the current pending `timeout tasks` and will not continue to flush any new spawned `timeout tasks` when flushing. 
So in the case above, after `flushOnlyPendingTimeout()`, there maybe still some timeout pendings (since they are newly spawned)

IgorMinar commented on packages/zone.js/lib/testing/fake-async.ts:
```diff
@@ -137,14 +137,85 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
+   *
+   * @experimental
+   */
+  function removeAllTimers(): void {
```
wouldn't it be better to call this one `cancelAllTimers` or `cancelAllScheduled` (see my general note about "timers" at the top)?

JiaLiPassion replied:
Got it, I will check it.

IgorMinar reviewed (commented):
Overall this looks quite good. I think that many of these APIs are quite useful.

I think that in order to land this, we'll need to improve the docs quite a bit and clarify what each of this utility function does.

One general thing I'd like to bring up is that it's a bit surprising that we use the "timers" terminology in a way where it includes not only `setTimeout` tasks, but also `setInterval` tasks and promise microtasks. Is this intentional and is it already ingrained in Zone.js APIs in a way where it doesn't make sense to correct it now? If it's too late to try to correct the terminology, then we should make an extra effort to explicitly spell out all of these categories of scheduled work so that the developers are not surprised.

IgorMinar commented:
btw the examples and usage notes in the PR description could also be incorporated into the inlined docs - they are quite helpful!

JiaLiPassion commented:
@IgorMinar , thank you for the review.

And I believe `fakeAsync` is handling not only `setTimeout` but also `setInterval` and `Promise` is intentional, so the following use cases can also be tested in the `sync` way. And we do have the public APIs so I believe the developers are aware of the features, https://angular.io/api/core/testing/discardPeriodicTasks

```
it('fakeAsync with setInterval', fakeAsync(() => {
  let num = 1;
  const id = setInterval(() => {
    num ++;
  }, 100);
  tick(100);
  expect(num).toBe(1);
  tick(100);
  expect(num).toBe(2);
  discardPeriodicTasks(); // Discard all remaining periodic tasks.
}));

it('fakeAsync with Promise', fakeAsync(() => {
  fixture.componentInstance.data = 'changed';
  fixture.detectChanges();
  fixture.whenStable(() => {
    expect(getText()).toEqual('changed');
  });
  tick();
}));
```

jelbourn commented:
While looking at this, I started thinking: these APIs are exported through angular/core/testing, but they depend on zone.js. Since optional zone.js is on the horizon, would it make sense to consider a different home for these APIs?
(would have to apply to e.g. `fakeAsync` as well in the long term)

jelbourn commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -156,3 +156,129 @@ export function flushMicrotasks(): void {
+/**
+ * Set fake System time.
+ *
+ * In `fakeAsync()`, the `Date.now()` is patched, so it
```
This mentions `Date.now()`- does this also apply to `(new Date()).getTime()`? If so, I would word these comments something like
```markdown
Sets the fake system time used for fakeAsync tests.

When using `fakeAsync()`, APIs that use the native system time, such as `Date.now()`, are
patched to use fake time values. This fake time advances only when explicitly calling `tick()`.
```
When it comes to method descriptions, it's generally better to start with the description of what the function does, followed by any necessary context. 

(similar for other methods)

JiaLiPassion replied:
Got it, thanks, I will update the comment here.

jelbourn replied:
There's still some grammatical issue with the descriptions. I would use the wording
```markdown
Sets the fake system time used for fakeAsync tests.

When using `fakeAsync()`, APIs that use the native system time, such as `Date.now()`, are
patched to use fake time values. This fake time advances only when explicitly calling `tick()`
```
* Provides context on what the fake system time is used for. 
* I think it's valuable to say "such as `Date.now()` since this also impacts `(new Date()).getTime()`.
* Mentions that fake time normally advances through `tick`, with the method being another alternative.

JiaLiPassion commented:
@jelbourn, yeah, these methods depends on `zone.js`, but in my understanding, the role of `zone.js` in `core/testing` is different with the role in `core/ngzone`, since `zone.js` in `core/testing` is just to manipulate the `async` APIs and make them easy to test, so it has nothing to do with the `change detection`, so it will also work even in the `zoneless` application.
So for `testing`, `zone.js` just another `timer` manipulation library like `Sinon.JS`, and `zone.js` is just a library to make the `mock timer` work.

JiaLiPassion committed:
fix(zone.js): fix jest setSystemTime implementation

Fix zone.js wrong implementation for `jest.setSystemTime()`, the current implementation is
forward `jest.setSystemTime()` to `fakeAsyncSpec.setFakeBaseSystemTime()`, and the `Date.now()`
returns the `fakeBaseSystemTime + currentTick` which is not what `jest.setSystemTime()` expected.
This commits fix the this issue.

jessicajaniuk commented:
Closing this since there's no activity on it, but feel free to re-open if you want to get it landed. Thanks, @JiaLiPassion.

JiaLiPassion commented:
@jessicajaniuk , sorry I forgot about this one, I am reworking on this one, so I will reopen it for now.

pkozlowski-opensource commented:
Hey @JiaLiPassion - looks like there was no movement on this PR for the last 5 months and this PR is still in the draft state. We are trying to keep the PR queue small so I'm going to close this PR for now. Feel free to open a PR when you've got it ready for review.
